extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityPolicy.java

```diff
@@ -1,30 +1,74 @@
 package io.quarkus.vertx.http.runtime.security;
 
 import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
 
 import io.quarkus.security.identity.SecurityIdentity;
-import io.vertx.core.http.HttpServerRequest;
 import io.vertx.ext.web.RoutingContext;
 
+/**
+ * A HTTP Security policy, that controls which requests are allowed to proceeed.
+ *
+ * There are two different ways these policies can be installed. The easiest is to just create a CDI bean, in which
+ * case the policy will be invoked on every request.
+ *
+ * Alternatively HttpSecurityPolicyBuildItem can be used to create a named policy. This policy can then be referenced
+ * in the application.properties path matching rules, which allows this policy to be applied to specific requests.
+ */
 public interface HttpSecurityPolicy {
 
-    CompletionStage<CheckResult> checkPermission(HttpServerRequest request, SecurityIdentity identity);
+    CompletionStage<CheckResult> checkPermission(RoutingContext request, SecurityIdentity identity,
+            AuthorizationRequestContext requestContext);
 
     /**
      * The results of a permission check
      */
-    enum CheckResult {
+    class CheckResult {
+
+        public static CheckResult DENY = new CheckResult(false);
+        public static CheckResult PERMIT = new CheckResult(true);
+
         /**
-         * If this is returned then the request is allowed. All permission checkers must permit a request for it
-         * to proceed.
+         * If this check was sucessful
          */
-        PERMIT,
+        private final boolean permitted;
+
         /**
-         * Denies the request. If the {@link SecurityIdentity} represents the anonymous user then
-         * {@link HttpAuthenticationMechanism#sendChallenge(RoutingContext)} will be invoked, otherwise
-         * a 403 forbidden error code will be returned.
+         * The new security identity, this allows the policy to add additional context
+         * information to the identity. If this is null no change is made
          */
-        DENY
+        private final SecurityIdentity augmentedIdentity;
+
+        public CheckResult(boolean permitted) {
+            this.permitted = permitted;
+            this.augmentedIdentity = null;
+        }
+
+        public CheckResult(boolean permitted, SecurityIdentity augmentedIdentity) {
+            this.permitted = permitted;
+            this.augmentedIdentity = augmentedIdentity;
+        }
+
+        public boolean isPermitted() {
+            return permitted;
+        }
+
+        public SecurityIdentity getAugmentedIdentity() {
+            return augmentedIdentity;
+        }
+    }
+
+    /**
+     * A context object that can be used to run blocking tasks
+     * <p>
+     * Blocking identity providers should used this context object to run blocking tasks, to prevent excessive and
+     * unnecessary delegation to thread pools
+     */
+    interface AuthorizationRequestContext {
```

Comment:
@stuartwdouglas Can `AuthorizationRequestContext` be optionally injected when needed into `HttpSecurityPolicy` implementations ? May be `HttpAuthorizer`'s `CONTEXT` can be CDI-produced ? So that the non-blocking implementations won't see it in the signature.
Or may be simpler, just renamed to something like `BlockingAuthorizationRequestContext` ?

Reply:
It may end up eventually containing more contextual information

Reply:
OK